Add some assertions to heap_stats.

damiendoligez opened the pull request:
In debug mode, check the order of free list blocks and heap chunks, and the total free list size.

damiendoligez committed:
in debug mode, check the order of free list blocks and heap chunks, as well as the total free list size

dra27 commented:
LGTM - that's extra hardening for the fix in #8585 and also in #1483 and finally closes down #1311

dra27 commented:
It's of course less good that the assertions are failing in various places!
```
tests/typing-misc/typecore_empty_polyvariant_error/ocamlc.byte/ocamlc.byte.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/misc/taku/ocamlc.byte/taku.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/boyer/ocamlc.byte/boyer.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/ephetest3/ocamlc.byte/ephetest3.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/ephetest3/ocamlopt.byte/ephetest3.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/nucleic/ocamlc.byte/nucleic.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/sorts/ocamlc.byte/sorts.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/sorts/ocamlopt.byte/sorts.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/ephetest/ocamlc.byte/ephetest.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc/ephetest/ocamlopt.byte/ephetest.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/basic-more/bounds/ocamlc.byte/bounds.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/basic-more/bounds/ocamlopt.byte/bounds.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray/bigarrays/ocamlc.byte/bigarrays.byte.output: 1... 2... 3... 4... 5... 6... 7... 8...file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray/bigarrays/ocamlopt.byte/bigarrays.opt.output: 1... 2... 3... 4... 5... 6... 7... 8...file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray/weak_bigarray/ocamlc.byte/weak_bigarray.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray/weak_bigarray/ocamlopt.byte/weak_bigarray.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray/fftba/ocamlc.byte/fftba.byte.output:16384... file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/tool-ocaml/t350-heapcheck/ocaml/ocaml.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/regression/pr5233/pr5233/ocamlc.byte/pr5233.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/regression/pr5233/pr5233/ocamlopt.byte/pr5233.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/regression/pr5757/pr5757/ocamlc.byte/pr5757.byte.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/regression/pr5757/pr5757/ocamlopt.byte/pr5757.opt.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/regression/pr3612/pr3612/ocamlc.byte/pr3612.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/regression/pr3612/pr3612/ocamlopt.byte/pr3612.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/tool-toplevel/pr7751/ocaml/ocaml.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/basic/arrays/ocamlc.byte/arrays.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/basic/arrays/ocamlopt.byte/arrays.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-systhreads/testpreempt/ocamlopt.byte/testpreempt.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-systhreads/threadsigmask/ocamlc.byte/threadsigmask.byte.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/lib-systhreads/threadsigmask/ocamlopt.byte/threadsigmask.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/ephe-c-api/test/ocamlc.byte/test.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/ephe-c-api/test/ocamlopt.byte/test.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/parsing/broken_invariants/ocamlc.byte/ocamlc.byte.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/lib-unix/common/getaddrinfo/ocamlc.byte/getaddrinfo.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-unix/common/getaddrinfo/ocamlopt.byte/getaddrinfo.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-bigarray-file/mapfile/ocamlopt.byte/mapfile.opt.output: 1...file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/self-contained-toplevel/main/ocamlc.byte/ocamlc.byte.output:file gc_ctrl.c; line 223 ### Assertion failed: Chunk_next (chunk) == NULL || Chunk_next (chunk) > chunk
tests/lib-set/testmap/ocamlopt.byte/testmap.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lib-set/testset/ocamlopt.byte/testset.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/gc-roots/globroots/ocamlc.byte/globroots.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/gc-roots/globroots/ocamlopt.byte/globroots.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/match-exception/streams/ocamlc.byte/streams.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/match-exception/streams/ocamlopt.byte/streams.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lazy/lazy1/ocamlc.byte/lazy1.byte.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/lazy/lazy1/ocamlopt.byte/lazy1.opt.output:file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
tests/misc-unsafe/fft/ocamlc.byte/fft.byte.output:32768... file gc_ctrl.c; line 233 ### Assertion failed: free_words == caml_fl_cur_wsz
```

damiendoligez committed:
do_compaction is called in two places, one of which breaks some of the
invariants added as assertions to heap_stats/caml_heap_check. We just
disable the call to heap_check in this case.

xavierleroy commented:
@damiendoligez: do you really want to keep this PR alive?  will it be relevant for OCaml 5.00 after the big GC rewrite?

damiendoligez commented:
No, this is not relevant to multicore.
